include/dmlc/intrusive_ptr.h

```diff
@@ -0,0 +1,259 @@
+  std::int32_t Dec() noexcept { return --count_; }
+  bool IsZero() const { return count_ == 0; }
+
+ public:
+  IntrusivePtrCell() noexcept : count_{0} {}
+  int32_t Count() const { return count_; }
+};
+
+/*!
+ * \brief User defined function for returing embedded reference count.
+ */
+template <typename T> IntrusivePtrCell &IntrusivePtrRefCount(T const *ptr) noexcept;
+
+/*!
+ * \brief Implementation of Intrusive Pointer.
+ *
+ * \code
+ *   struct RefCountedData {
+ *     mutable class IntrusivePtrCell ref;
+ *     float data { 0 };
+ *
+ *     friend IntrusivePtrCell &
+ *     IntrusivePtrRefCount(RefCountedData const *t) noexcept {
+ *       return t->ref;
+ *     }
+ *   };
+ *
+ *   auto p = make_intrusive<RefCountedData>();
+ *
+ * \endcode
+ */
+template <typename T> class IntrusivePtr {
+ private:
+  void Incref(T *ptr) {
+    if (ptr) {
+      IntrusivePtrRefCount(ptr).Inc();
+    }
+  }
+  void Decref(T *ptr) {
+    if (ptr) {
+      if (IntrusivePtrRefCount(ptr).Dec() == 0) {
+        delete ptr;
+      }
+    }
+  }
+
+  /*!
+   * \brief Contruct an IntrusivePtr.
+   *
+   * \param p Raw pointer to object
+   */
+  explicit IntrusivePtr(T *p) : ptr_{p} {
+    if (ptr_) {
+      Incref(ptr_);
+    }
+  }
+
+ protected:
+  T *ptr_{nullptr};
+
+ public:
+  using element_type = T;  // NOLINT
+  struct Hash {
+    std::size_t operator()(IntrusivePtr<element_type> const &ptr) const noexcept {
+      return std::hash<element_type *>()(ptr.get());
+    }
+  };
+
+  template <typename U, typename... Args>
+  friend IntrusivePtr<U> make_intrusive(Args&&... args);  // NOLINT
+  template <typename Base, typename Sub,
+            typename... Args>
+  friend IntrusivePtr<Base> make_intrusive_base(Args &&... args); // NOLINT
+  template <typename X, typename Y>
+  friend IntrusivePtr<X>
+  dynamic_pointer_cast(IntrusivePtr<Y> const &p); // NOLINT
+
+  IntrusivePtr() noexcept = default;
+  IntrusivePtr(IntrusivePtr const &that) : ptr_{that.ptr_} { Incref(ptr_); }
+
+  template <typename U>
+  explicit IntrusivePtr(IntrusivePtr<U> const &that) : ptr_{that.ptr_} {
+    Incref(ptr_);
+  }
+
+  IntrusivePtr(IntrusivePtr &&that) : ptr_{that.ptr_} { that.ptr_ = nullptr; }
+
+  template <typename U>
+  explicit IntrusivePtr(IntrusivePtr<U> &&that) : ptr_{that.ptr_} {
+    that.ptr_ = nullptr;
+  }
+
+  ~IntrusivePtr() { Decref(ptr_); }
+
+  IntrusivePtr &operator=(element_type *that) {
+    *this = IntrusivePtr(that);
+    return *this;
+  }
+  IntrusivePtr &operator=(IntrusivePtr const &that) {
+    auto copy = that;
+    std::swap(copy, *this);
+    return *this;
+  }
+  template <typename U> IntrusivePtr &operator=(IntrusivePtr<U> const &that) {
+    IntrusivePtr<element_type> copy{that.get()};
+    std::swap(*this, copy);
+    return *this;
+  }
+
+  IntrusivePtr &operator=(IntrusivePtr &&that) {
+    std::swap(ptr_, that.ptr_);
+    return *this;
+  }
+  template <typename U> IntrusivePtr &operator=(IntrusivePtr<U> &&that) {
```

Comment:
This is unsafe, unless there is a clear relation between U and T.

Reply:
Removed.